File: src/aiusage/render.py
import datetime as dt
import re
import unicodedata

from .i18n import tr
from .models import Availability, ProviderUsage
from .timezones import convert, from_epoch, offset_label

ANSI = re.compile(r"\x1b\[[0-9;]*m")
MIN_CELL_WIDTH = 24
CELL_GAP = 2
RESET_STYLE = "\x1b[0m"
THEME_STYLE = {
    "white": {"normal": "\x1b[37m", "strong": "\x1b[1;97m", "muted": "\x1b[90m"},
    "green": {"normal": "\x1b[32m", "strong": "\x1b[1;92m", "muted": "\x1b[32m"},
}


def visible_len(value):
    return sum(_char_width(char) for char in ANSI.sub("", value))


def _char_width(char):
    if unicodedata.combining(char) or unicodedata.category(char).startswith("C"):
        return 0
    return 2 if unicodedata.east_asian_width(char) in ("W", "F") else 1


def _fit(value, width):
    result = []
    used = 0
    for char in value:
        size = _char_width(char)
        if used + size > max(0, width):
            break
        result.append(char)
        used += size
    return "".join(result)


def _pad(value, width):
    value = _fit(value, width)
    return value + " " * max(0, width - visible_len(value))


def reset_text(epoch, language="en", timezone="system"):
    """Convert and format a reset epoch in the selected display timezone."""
    if not isinstance(epoch, (int, float)):
        return "?"
    value = from_epoch(epoch, timezone)
    zone = offset_label(value)
    if language == "zh":
        return f"{value.month}月{value.day:02d}日 {value:%H:%M} {zone}"
    return f"{value:%b %d %H:%M} {zone}"


def system_text(language="en", timezone="system", now=None):
    value = convert(now or dt.datetime.now(dt.timezone.utc), timezone)
    zone = offset_label(value)
    return _label_value(tr(language, "system"), f"{value:%Y-%m-%d %H:%M:%S} {zone}", language)


def _label_value(label, value, language):
    return f"{label}：{value}" if language == "zh" else f"{label}: {value}"


def column_count(count, width):
    if count <= 2:
        return 1
    capacity = max(1, min(3, (width + CELL_GAP) // (MIN_CELL_WIDTH + CELL_GAP)))
    if count == 4 and capacity >= 2:
        return 2
    return min(count, capacity)


def _provider_lines(provider: ProviderUsage, width: int, language: str, timezone: str):
    suffix = f"  ! {tr(language, 'stale')}" if provider.stale else ""
    lines = [provider.name.upper() + suffix]
    if provider.availability != Availability.AVAILABLE or not provider.windows:
        lines.append(tr(language, provider.availability.value))
        return [_fit(line, width) for line in lines]

    bar_width = max(3, min(10, width - 17))
    for window in provider.windows[:2]:
        remaining = max(0, min(100, int(window.remaining_percent)))
        fill = round(remaining * bar_width / 100)
        bar = "█" * fill + "░" * (bar_width - fill)
        label = "Weekly" if window.label == "Week" and language == "en" else window.label
        lines.append(f"{_fit(label, 6):<6} {bar} {remaining:>3}% {tr(language, 'left')}")
        reset_value = reset_text(window.reset_at, language, timezone)
        reset_line = _label_value(tr(language, "reset"), reset_value, language)
        # In narrow grids preserve the complete instant and UTC offset; the
        # repeated Reset label is less important than an unambiguous timestamp.
        lines.append(reset_line if visible_len(reset_line) <= width else reset_value)
    return [_fit(line, width) for line in lines]


def _top_border(width, title):
    inner = max(0, width - 2)
    token = _fit(f" {title} ", inner)
    remaining = max(0, inner - visible_len(token))
    left = remaining // 2
    return "┌" + "─" * left + token + "─" * (remaining - left) + ("┐" if width > 1 else "")


def _outer_line(value, width):
    if width < 2:
        return _fit(value, width)
    inner = width - 2
    return "│" + _pad(value, inner) + "│"


def _position(lines, terminal_width, terminal_height, position):
    block_width = max((visible_len(line) for line in lines), default=0)
    block_height = len(lines)
    vertical, horizontal = ("center", "center") if position == "center" else position.split("-", 1)
    left = {
        "left": 0,
        "center": max(0, (terminal_width - block_width) // 2),
        "right": max(0, terminal_width - block_width),
    }[horizontal]
    top = {
        "top": 0,
        "center": max(0, (terminal_height - block_height) // 2),
        "bottom": max(0, terminal_height - block_height),
    }[vertical]
    return [""] * top + [_fit(" " * left + line, terminal_width) for line in lines]


def _style(lines, kinds, theme, color):
    if not color:
        return lines
    palette = THEME_STYLE.get(theme, THEME_STYLE["white"])
    return [
        f"{palette.get(kind, palette['normal'])}{line}{RESET_STYLE}" if line else line
        for line, kind in zip(lines, kinds)
    ]


def dashboard(
    width,
    height,
    providers,
    updated,
    language="en",
    position="center",
    demo=False,
    theme="white",
    color=False,
    timezone="system",
):
    width, height = max(1, width), max(1, height)
    title = f"AI USAGE [{tr(language, 'demo')}]" if demo else "AI USAGE"
    if width < 4 or height < 3:
        return _position([_fit(title, width)][:height], width, height, position)

    help_text = tr(language, "help")
    compact_help = tr(language, "help_compact")
    updated_at = convert(updated, timezone) if updated else None
    stamp = updated_at.strftime("%H:%M:%S") if updated_at else "--:--:--"
    update_text = _label_value(tr(language, "updated"), stamp, language)

    content_capacity = max(1, width - 4)
    columns = column_count(len(providers), content_capacity)
    preferred_cell = {1: 40, 2: 27, 3: 24}[columns]
    cell_width = min(preferred_cell, max(1, (content_capacity - CELL_GAP * (columns - 1)) // columns))
    grid_width = cell_width * columns + CELL_GAP * (columns - 1)
    desired_outer = max(grid_width + 4, visible_len(compact_help) + 4, visible_len(system_text(language, timezone)) + 4)
    outer_width = min(width, desired_outer)
    inner_width = outer_width - 2
    content_width = max(1, inner_width - 2)
    cell_width = min(cell_width, max(1, (content_width - CELL_GAP * (columns - 1)) // columns))
    grid_width = cell_width * columns + CELL_GAP * (columns - 1)

    blocks = [_provider_lines(provider, cell_width, language, timezone) for provider in providers]

    content = [""]
    content_kinds = ["normal"]
    rows = (len(blocks) + columns - 1) // columns
    for row_index in range(rows):
        group = blocks[row_index * columns:(row_index + 1) * columns]
        row_height = max((len(block) for block in group), default=1)
        group = [block + [""] * (row_height - len(block)) for block in group]
        group_width = len(group) * cell_width + max(0, len(group) - 1) * CELL_GAP
        grid_left = max(0, (content_width - group_width) // 2)
        for line_index in range(row_height):
            value = (" " * CELL_GAP).join(_pad(block[line_index], cell_width) for block in group)
            content.append(" " + " " * grid_left + value)
            content_kinds.append("strong" if line_index == 0 else "normal")
        if row_index < rows - 1:
            content.append("")
            content_kinds.append("normal")

    chosen_help = help_text if visible_len(help_text) <= content_width - 1 else compact_help
    content.extend(["", " " + system_text(language, timezone), " " + update_text, "", " " + chosen_help, ""])
    content_kinds.extend(["normal", "normal", "muted", "normal", "muted", "normal"])

    max_content_lines = max(0, height - 2)
    if len(content) > max_content_lines:
        content = content[:max_content_lines]
        content_kinds = content_kinds[:max_content_lines]

    raw_lines = [_top_border(outer_width, title)]
    kinds = ["strong"]
    for value, kind in zip(content, content_kinds):
        raw_lines.append(_outer_line(_fit(value, inner_width), outer_width))
        kinds.append(kind)
    raw_lines.append("└" + "─" * (outer_width - 2) + "┘")
    kinds.append("normal")

    positioned = _position(raw_lines[:height], width, height, position)
    top_padding = len(positioned) - len(raw_lines[:height])
    styled = _style(positioned[top_padding:], kinds[: len(positioned) - top_padding], theme, color)
    return positioned[:top_padding] + styled


def selector(width, height, registry, enabled, cursor, language, theme="white", color=False):
    box_width = min(58, max(30, width - 4))
    inner = box_width - 2
    lines = ["┌" + "─" * inner + "┐"]
    title = tr(language, "providers")
    lines.append("│ " + _pad(title, inner - 2) + " │")
    for index, (key, adapter) in enumerate(registry.items()):
        mark = "x" if key in enabled else " "
        pointer = "›" if index == cursor else " "
        order = enabled.index(key) + 1 if key in enabled else 0
        value = f"{pointer} [{mark}] {adapter.name}" + (f"  {order}" if order else "")
        lines.append("│ " + _pad(value, inner - 2) + " │")
    help_value = _fit(tr(language, "select_help"), inner - 2)
    lines.append("│ " + _pad(help_value, inner - 2) + " │")
    lines.append("└" + "─" * inner + "┘")
    positioned = _position(lines, width, height, "center")
    top_padding = len(positioned) - len(lines)
    styled = _style(positioned[top_padding:], ["normal"] * len(lines), theme, color)
    return positioned[:top_padding] + styled


def timezone_selector(width, height, options, cursor, language, theme="white", color=False):
    box_width = min(60, max(34, width - 4))
    inner = box_width - 2
    lines = [_top_border(box_width, tr(language, "timezone"))]
    for index, value in enumerate(options):
        pointer = "›" if index == cursor else " "
        if value == "system":
            label = tr(language, "system_zone")
        elif index == len(options) - 1:
            label = f"{tr(language, 'custom_zone')}: {value}"
        else:
            label = value
        lines.append("│ " + _pad(f"{pointer} {label}", inner - 2) + " │")
    lines.append("│ " + _pad(_fit(tr(language, "timezone_help"), inner - 2), inner - 2) + " │")
    lines.append("└" + "─" * inner + "┘")
    positioned = _position(lines, width, height, "center")
    top_padding = len(positioned) - len(lines)
    styled = _style(positioned[top_padding:], ["normal"] * len(lines), theme, color)
    return positioned[:top_padding] + styled

File: rust/src/render.rs
//! Pure frame assembly; explicit clock makes byte-level compatibility testable.
use chrono::Datelike;
use unicode_width::UnicodeWidthChar;
use crate::{config::Config, i18n::tr, models::{Availability, ProviderUsage}, timezones, PROVIDERS};

pub fn char_width(c:char)->usize {c.width().unwrap_or(0)}
pub fn visible_len(text:&str)->usize {
    let mut chars=text.chars().peekable();
    let mut count=0;
    while let Some(c)=chars.next() {
        if c=='\x1b' && chars.peek()==Some(&'[') {
            let saved=chars.clone(); chars.next();
            while chars.peek().is_some_and(|c|c.is_ascii_digit()||*c==';') {chars.next();}
            if chars.peek()==Some(&'m') {chars.next();continue;}
            chars=saved;
        }
        count+=char_width(c);
    }
    count
}
pub fn fit(text:&str,width:usize)->String {
    let mut used=0;
    text.chars().take_while(|c| {used+=char_width(*c);used<=width}).collect()
}
pub fn pad(text:&str,width:usize)->String {
    let value=fit(text,width);
    format!("{value}{}"," ".repeat(width.saturating_sub(visible_len(&value))))
}
pub fn label_value(label:&str,value:&str,language:&str)->String {
    format!("{label}{}{value}",if language=="zh" {"："} else {": "})
}
pub fn reset_text(epoch:Option<f64>,language:&str,timezone:&str)->String {
    let Some(value)=epoch.and_then(|e|timezones::from_epoch(e,timezone).ok()) else {return "?".into()};
    let stamp=if language=="zh" {format!("{}月{:02}日 {}",value.month(),value.day(),value.format("%H:%M"))}
        else {value.format("%b %d %H:%M").to_string()};
    format!("{stamp} {}",timezones::label_for(&value))
}
pub fn system_text(language:&str,timezone:&str,now:f64)->String {
    let value=timezones::from_epoch(now,timezone).expect("validated timezone/clock");
    label_value(tr(language,"system"),&format!("{} {}",value.format("%Y-%m-%d %H:%M:%S"),timezones::label_for(&value)),language)
}
pub fn column_count(count:usize,width:usize)->usize {
    if count<=2 {return 1;}
    let capacity=((width+2)/26).clamp(1,3);
    if count==4 && capacity>=2 {2} else {count.min(capacity)}
}
fn provider_lines(provider:&ProviderUsage,width:usize,language:&str,timezone:&str)->Vec<String> {
    let mut lines=vec![format!("{}{}",provider.name.to_uppercase(),if provider.stale {format!("  ! {}",tr(language,"stale"))} else {String::new()})];
    if provider.availability!=Availability::Available || provider.windows.is_empty() {
        let key=match provider.availability {Availability::Available=>"available",Availability::NotInstalled=>"not_installed",Availability::Unavailable=>"unavailable",Availability::NotSupported=>"not_supported"};
        lines.push(tr(language,key).into());
    } else {
        let bar_width=width.saturating_sub(17).clamp(3,10);
        for window in provider.windows.iter().take(2) {
            let remaining=window.remaining_percent.clamp(0,100);
            let fill=(remaining as f64*bar_width as f64/100.0).round_ties_even() as usize;
            let label=if window.label=="Week" && language=="en" {"Weekly"} else {&window.label};
            let label=fit(label,6);
            // Python's <6 pads by codepoints here, not terminal cells.
            let label=format!("{label}{}"," ".repeat(6usize.saturating_sub(label.chars().count())));
            lines.push(format!("{label} {}{} {remaining:>3}% {}","█".repeat(fill),"░".repeat(bar_width-fill),tr(language,"left")));
            let value=reset_text(window.reset_at,language,timezone);
            let full=label_value(tr(language,"reset"),&value,language);
            lines.push(if visible_len(&full)<=width {full} else {value});
        }
    }
    lines.into_iter().map(|s|fit(&s,width)).collect()
}
fn top_border(width:usize,title:&str)->String {
    let inner=width.saturating_sub(2);
    let token=fit(&format!(" {title} "),inner);
    let remaining=inner.saturating_sub(visible_len(&token));
    format!("┌{}{token}{}{}","─".repeat(remaining/2),"─".repeat(remaining-remaining/2),if width>1 {"┐"} else {""})
}
fn outer_line(text:&str,width:usize)->String {
    if width<2 {fit(text,width)} else {format!("│{}│",pad(text,width-2))}
}
fn position(lines:Vec<String>,width:usize,height:usize,pos:&str)->Vec<String> {
    let block_width=lines.iter().map(|s|visible_len(s)).max().unwrap_or(0);
    let (vertical,horizontal)=if pos=="center" {("center","center")} else {pos.split_once('-').unwrap_or(("center","center"))};
    let left=match horizontal {"left"=>0,"right"=>width.saturating_sub(block_width),_=>width.saturating_sub(block_width)/2};
    let top=match vertical {"top"=>0,"bottom"=>height.saturating_sub(lines.len()),_=>height.saturating_sub(lines.len())/2};
    let mut result=vec![String::new();top];
    result.extend(lines.iter().map(|line|fit(&format!("{}{line}"," ".repeat(left)),width)));
    result
}
fn style(lines:&mut [String],kinds:&[&str],theme:&str,color:bool) {
    if !color {return;}
    for (line,kind) in lines.iter_mut().zip(kinds) {
        if line.is_empty() {continue;}
        let code=match (theme,*kind) {("green","strong")=>"1;92",("green",_)=>"32",(_,"strong")=>"1;97",(_,"muted")=>"90",_=>"37"};
        *line=format!("\x1b[{code}m{line}\x1b[0m");
    }
}

pub struct Frame<'a> {
    pub width:usize, pub height:usize, pub providers:&'a [ProviderUsage],
    pub updated:Option<f64>, pub now:f64, pub config:&'a Config,
    pub demo:bool, pub color:bool, pub notice:Option<&'a str>,
}
pub fn dashboard(frame:Frame<'_>)->Vec<String> {
    let Frame{width,height,providers,updated,now,config:cfg,demo,color,notice}=frame;
    let (width,height)=(width.max(1),height.max(1));
    let language=&cfg.language;
    let title=if demo {format!("AI USAGE [{}]",tr(language,"demo"))} else {"AI USAGE".into()};
    if width<4 || height<3 {return position(vec![fit(&title,width)],width,height,&cfg.position);}
    let stamp=updated.and_then(|e|timezones::from_epoch(e,&cfg.timezone).ok()).map(|v|v.format("%H:%M:%S").to_string()).unwrap_or("--:--:--".into());
    let update=label_value(tr(language,"updated"),&stamp,language);
    let system=system_text(language,&cfg.timezone,now);
    let capacity=width.saturating_sub(4).max(1);
    let columns=column_count(providers.len(),capacity);
    let preferred=[0,40,27,24][columns];
    let cell=preferred.min(capacity.saturating_sub(2*(columns-1)).checked_div(columns).unwrap().max(1));
    let grid=cell*columns+2*(columns-1);
    let outer=width.min((grid+4).max(visible_len(tr(language,"help_compact"))+4).max(visible_len(&system)+4));
    let inner=outer-2;
    let content_width=inner.saturating_sub(2).max(1);
    let cell=cell.min((content_width.saturating_sub(2*(columns-1))/columns).max(1));
    let blocks:Vec<_>=providers.iter().map(|p|provider_lines(p,cell,language,&cfg.timezone)).collect();
    let mut content=vec![String::new()];
    let mut kinds=vec!["normal"];
    for (row_index,group) in blocks.chunks(columns).enumerate() {
        let row_height=group.iter().map(Vec::len).max().unwrap_or(1);
        let group_width=group.len()*cell+group.len().saturating_sub(1)*2;
        let left=content_width.saturating_sub(group_width)/2;
        for index in 0..row_height {
            let value=group.iter().map(|b|pad(b.get(index).map(String::as_str).unwrap_or(""),cell)).collect::<Vec<_>>().join("  ");
            content.push(format!(" {}{value}"," ".repeat(left)));
            kinds.push(if index==0 {"strong"} else {"normal"});
        }
        if (row_index+1)*columns<blocks.len() {content.push(String::new());kinds.push("normal");}
    }
    let help=tr(language,if visible_len(tr(language,"help"))<=content_width.saturating_sub(1) {"help"} else {"help_compact"});
    content.extend([String::new(),format!(" {system}"),format!(" {update}")]);
    kinds.extend(["normal","normal","muted"]);
    if let Some(notice)=notice {content.push(format!(" {notice}"));kinds.push("strong");}
    content.extend([String::new(),format!(" {help}"),String::new()]);kinds.extend(["normal","muted","normal"]);
    content.truncate(height-2);kinds.truncate(height-2);
    let mut raw=vec![top_border(outer,&title)];
    raw.extend(content.iter().map(|s|outer_line(&fit(s,inner),outer)));
    raw.push(format!("└{}┘","─".repeat(outer-2)));
    kinds.insert(0,"strong");kinds.push("normal");
    raw.truncate(height);
    let count=raw.len();
    let mut positioned=position(raw,width,height,&cfg.position);
    let top=positioned.len()-count;
    style(&mut positioned[top..],&kinds,&cfg.theme,color);
    positioned
}

pub fn selector(width:usize,height:usize,enabled:&[String],cursor:usize,cfg:&Config,color:bool,discovery:&std::collections::HashMap<String,String>)->Vec<String> {
    let box_width=width.saturating_sub(4).clamp(30,58);
    let inner=box_width-2;
    let row=|s:&str|format!("│ {} │",pad(s,inner-2));
    let mut lines=vec![format!("┌{}┐","─".repeat(inner)),row(tr(&cfg.language,"providers"))];
    for (index,(key,name)) in PROVIDERS.iter().enumerate() {
        let order=enabled.iter().position(|k|k==key).map(|n|n+1);
        let mut value=format!("{} [{}] {name}{}",if index==cursor {"›"} else {" "},if order.is_some() {"x"} else {" "},order.map(|n|format!("  {n}")).unwrap_or_default());
        if let Some(reason)=discovery.get(*key) {
            let available=(inner-3).saturating_sub(visible_len(&value));
            if available>=5 {value.push_str(&format!("  {}",fit(tr(&cfg.language,reason),available-2)));}
        }
        lines.push(row(&value));
    }
    lines.push(row(&fit(tr(&cfg.language,"select_help"),inner-2)));
    lines.push(format!("└{}┘","─".repeat(inner)));
    selection_position(lines,width,height,cfg,color)
}
fn selection_position(lines:Vec<String>,width:usize,height:usize,cfg:&Config,color:bool)->Vec<String> {
    let count=lines.len();let mut lines=position(lines,width,height,"center");let top=lines.len()-count;
    style(&mut lines[top..],&vec!["normal";count],&cfg.theme,color);lines
}
pub fn timezone_selector(width:usize,height:usize,options:&[String],cursor:usize,cfg:&Config,color:bool)->Vec<String> {
    let box_width=width.saturating_sub(4).clamp(34,60);let inner=box_width-2;
    let row=|s:&str|format!("│ {} │",pad(s,inner-2));
    let mut lines=vec![top_border(box_width,tr(&cfg.language,"timezone"))];
    for (index,value) in options.iter().enumerate() {
        let label=if value=="system" {tr(&cfg.language,"system_zone").into()}
            else if index==options.len()-1 {format!("{}: {value}",tr(&cfg.language,"custom_zone"))} else {value.clone()};
        lines.push(row(&format!("{} {label}",if index==cursor {"›"} else {" "})));
    }
    lines.push(row(&fit(tr(&cfg.language,"timezone_help"),inner-2)));lines.push(format!("└{}┘","─".repeat(inner)));
    selection_position(lines,width,height,cfg,color)
}
